mava/components/training/losses.py

```diff
@@ -19,22 +19,68 @@
 from types import SimpleNamespace
 from typing import Any, Dict, List, Tuple, Type
 
+import chex
 import haiku as hk
 import jax
 import jax.numpy as jnp
 import rlax
+from chex import Array
 from haiku._src.basic import merge_leading_dims
 
 from mava.callbacks import Callback
 from mava.components import Component, training
 from mava.core_jax import SystemTrainer
 
 
+def clipped_surrogate_pg_loss(
+    prob_ratios_t: Array,
+    adv_t: Array,
+    epsilon: float,
+    mask: Array,
+    use_stop_gradient: bool = True,
+) -> Array:
+    """Computes the clipped surrogate policy gradient loss.
+
+    L_clipₜ(θ) = - min(rₜ(θ)Âₜ, clip(rₜ(θ), 1-ε, 1+ε)Âₜ)
+    Where rₜ(θ) = π_θ(aₜ| sₜ) / π_θ_old(aₜ| sₜ) and Âₜ are the advantages.
+    See Proximal Policy Optimization Algorithms, Schulman et al.:
+    https://arxiv.org/abs/1707.06347
+    -----------------------------------------------
+    This was copied from the RLAX packaged and only modifed to accept a mask
+    -----------------------------------------------
+
+    Args:
+        prob_ratios_t:
+            Ratio of action probabilities for actions a_t:
+            rₜ(θ) = π_θ(aₜ| sₜ) / π_θ_old(aₜ| sₜ)
+        adv_t:
+            the observed or estimated advantages from executing actions a_t.
+        epsilon:
+            Scalar value corresponding to how much to clip the objecctive.
+        mask:
+            Mask to apply for sequence padding.
+        use_stop_gradient:
+            bool indicating whether or not to apply stop gradient to
+            advantages.
+
+    Returns:
+        Loss whose gradient corresponds to a clipped surrogate policy gradient
+            update.
+    """
+    chex.assert_rank([prob_ratios_t, adv_t], [1, 1])
+    chex.assert_type([prob_ratios_t, adv_t], [float, float])
+
+    adv_t = jax.lax.select(use_stop_gradient, jax.lax.stop_gradient(adv_t), adv_t)
+    clipped_ratios_t = jnp.clip(prob_ratios_t, 1.0 - epsilon, 1.0 + epsilon)
+    clipped_objective = jnp.fmin(prob_ratios_t * adv_t, clipped_ratios_t * adv_t)
+
+    return -jnp.sum(clipped_objective * mask) / jnp.sum(mask)
```

Comment:
This could lead to an error that happen to us priorly, at some cases you may divide by zero `jnp.sum(mask)=0`(I know it may be rare but at least add a small factor next to it)

Reply:
Yep. I just pushed some updates for with and also added a flag for the sequence masking.

Reply:
Cool thanks :slightly_smiling_face: